File: example280126/src/main/java/ru/examle/edu/service/InvitationService.java
package ru.examle.edu.service;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import ru.examle.edu.dto.*;
import ru.examle.edu.entity.Invitation;
import ru.examle.edu.entity.Meeting;
import ru.examle.edu.entity.Person;
import ru.examle.edu.entity.Invitation.InvitationStatus;
import ru.examle.edu.exception.ResourceNotFoundException;
import ru.examle.edu.exception.ValidationException;
import ru.examle.edu.repository.InvitationRepository;
import ru.examle.edu.repository.MeetingRepository;
import ru.examle.edu.repository.PersonRepository;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class InvitationService {

    private final InvitationRepository invitationRepository;
    private final MeetingRepository meetingRepository;
    private final PersonRepository personRepository;

    public List<InvitationDto> getAllInvitations() {
        return invitationRepository.findAll().stream()
                .map(this::convertToDto)
                .collect(Collectors.toList());
    }

    public InvitationDto getInvitationById(Long id) {
        Invitation invitation = invitationRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Invitation not found with id: " + id));
        return convertToDto(invitation);
    }

    @Transactional
    public InvitationDto createInvitation(InvitationRequest request) {
        Meeting meeting = meetingRepository.findById(request.getMeetingId())
                .orElseThrow(() -> new ResourceNotFoundException("Meeting not found with id: " + request.getMeetingId()));

        Person person = personRepository.findById(request.getPersonId())
                .orElseThrow(() -> new ResourceNotFoundException("Person not found with id: " + request.getPersonId()));

        if (meeting.getOrganizer().getId().equals(person.getId())) {
            throw new ValidationException("Organizer cannot be invited to their own meeting");
        }

        invitationRepository.findByMeetingIdAndPersonId(meeting.getId(), person.getId())
                .ifPresent(inv -> {
                    throw new ValidationException("Invitation already exists for this meeting and person");
                });

        Invitation invitation = new Invitation();
        invitation.setMeeting(meeting);
        invitation.setPerson(person);
        invitation.setStatus(request.getStatus() != null ? request.getStatus() : InvitationStatus.PENDING);
        invitation.setCreatedAt(LocalDateTime.now());
        invitation.setUpdatedAt(LocalDateTime.now());

        invitation = invitationRepository.save(invitation);
        return convertToDto(invitation);
    }

    @Transactional
    public InvitationDto updateInvitationStatus(Long id, InvitationUpdateRequest request) {
        Invitation invitation = invitationRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Invitation not found with id: " + id));

        if (invitation.getStatus() == InvitationStatus.CANCELLED) {
            throw new ValidationException("Cannot update a cancelled invitation");
        }

        if (request.getStatus() == InvitationStatus.ACCEPTED ||
                request.getStatus() == InvitationStatus.DECLINED) {
            invitation.setResponseDate(LocalDateTime.now());
        }

        invitation.setStatus(request.getStatus());
        invitation.setUpdatedAt(LocalDateTime.now());
        invitation = invitationRepository.save(invitation);
        return convertToDto(invitation);
    }

    @Transactional
    public void deleteInvitation(Long id) {
        Invitation invitation = invitationRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Invitation not found with id: " + id));

        invitationRepository.delete(invitation);
    }

    public List<InvitationDto> getInvitationsByPerson(Long personId) {
        return invitationRepository.findByPersonId(personId).stream()
                .map(this::convertToDto)
                .collect(Collectors.toList());
    }

    public List<InvitationDto> getInvitationsByMeeting(Long meetingId) {
        return invitationRepository.findByMeetingId(meetingId).stream()
                .map(this::convertToDto)
                .collect(Collectors.toList());
    }

    private InvitationDto convertToDto(Invitation invitation) {
        InvitationDto dto = new InvitationDto();
        dto.setId(invitation.getId());
        dto.setStatus(invitation.getStatus());
        dto.setResponseDate(invitation.getResponseDate());
        dto.setCreatedAt(invitation.getCreatedAt());
        dto.setUpdatedAt(invitation.getUpdatedAt());

        if (invitation.getMeeting() != null) {
            MeetingSimpleDto meetingDto = new MeetingSimpleDto();
            meetingDto.setId(invitation.getMeeting().getId());
            meetingDto.setTitle(invitation.getMeeting().getTitle());
            meetingDto.setMeetingDate(invitation.getMeeting().getMeetingDate());
            meetingDto.setStartTime(invitation.getMeeting().getStartTime());
            meetingDto.setEndTime(invitation.getMeeting().getEndTime());
            meetingDto.setStatus(invitation.getMeeting().getStatus());
            dto.setMeeting(meetingDto);
        }

        if (invitation.getPerson() != null) {
            PersonSimpleDto personDto = new PersonSimpleDto();
            personDto.setId(invitation.getPerson().getId());
            personDto.setName(invitation.getPerson().getName());
            personDto.setEmail(invitation.getPerson().getEmail());
            dto.setPerson(personDto);
        }

        return dto;
    }
}

File: src/main/kotlin/ru/sicampus/bootcamp2026/service/InvitationService.kt
package ru.sicampus.bootcamp2026.service

import org.springframework.security.access.AccessDeniedException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import ru.sicampus.bootcamp2026.dto.InvitationCreateDto
import ru.sicampus.bootcamp2026.dto.InvitationResponseDto
import ru.sicampus.bootcamp2026.dto.InvitationUpdateDto
import ru.sicampus.bootcamp2026.entity.Invitation
import ru.sicampus.bootcamp2026.entity.InvitationStatus
import ru.sicampus.bootcamp2026.repository.InvitationRepository
import ru.sicampus.bootcamp2026.repository.MeetingRepository
import ru.sicampus.bootcamp2026.repository.UserRepository
import ru.sicampus.bootcamp2026.security.SecurityUtils
import java.util.*

@Service
@Transactional
class InvitationService(
    private val invitationRepository: InvitationRepository,
    private val meetingRepository: MeetingRepository,
    private val userRepository: UserRepository
) {
    fun getInvitations(
        meetingId: Long? = null,
        userId: Long? = null,
        status: InvitationStatus? = null
    ): List<InvitationResponseDto> {
        val currentUser = SecurityUtils.getCurrentUser(userRepository)
        val isAdmin = SecurityUtils.isAdmin(currentUser)
        
        userId?.let { 
            SecurityUtils.requireOwnershipOrAdmin(currentUser, it)
        }
        
        meetingId?.let {
            val meeting = meetingRepository.findById(it)
                .orElseThrow { NoSuchElementException("Meeting with id $it not found") }
            SecurityUtils.requireOwnershipOrAdmin(currentUser, meeting.organizer.id)
        }
        
        val effectiveUserId = if (isAdmin) userId else (userId ?: currentUser.id)
        
        val invitations = when {
            meetingId != null && effectiveUserId != null && status != null -> {
                invitationRepository.findByMeetingId(meetingId)
                    .filter { it.user.id == effectiveUserId && it.status == status }
            }
            meetingId != null && effectiveUserId != null -> {
                invitationRepository.findByMeetingIdAndUserId(meetingId, effectiveUserId)
            }
            meetingId != null && status != null -> {
                invitationRepository.findByMeetingId(meetingId)
                    .filter { it.status == status }
            }
            effectiveUserId != null && status != null -> {
                invitationRepository.findByUserId(effectiveUserId)
                    .filter { it.status == status }
            }
            meetingId != null -> {
                invitationRepository.findByMeetingId(meetingId)
            }
            effectiveUserId != null -> {
                invitationRepository.findByUserId(effectiveUserId)
            }
            status != null -> {
                if (isAdmin) {
                    invitationRepository.findByStatus(status)
                } else {
                    invitationRepository.findByUserId(currentUser.id)
                        .filter { it.status == status }
                }
            }
            else -> {
                if (isAdmin) {
                    invitationRepository.findAll()
                } else {
                    invitationRepository.findByUserId(currentUser.id)
                }
            }
        }
        
        return invitations.map { it.toResponseDto() }
    }

    fun getInvitationById(id: Long): InvitationResponseDto {
        val currentUser = SecurityUtils.getCurrentUser(userRepository)
        val invitation = invitationRepository.findById(id)
            .orElseThrow { NoSuchElementException("Invitation with id $id not found") }
        
        if (
            currentUser.role.name != "ADMIN" 
            && invitation.user.id != currentUser.id 
            && invitation.meeting.organizer.id != currentUser.id
        ) {
            throw AccessDeniedException("Access denied")
        }
        
        return invitation.toResponseDto()
    }

    fun createInvitation(dto: InvitationCreateDto): InvitationResponseDto {
        val currentUser = SecurityUtils.getCurrentUser(userRepository)
        val meeting = meetingRepository.findById(dto.meetingId)
            .orElseThrow { NoSuchElementException("Meeting with id ${dto.meetingId} not found") }
        
        SecurityUtils.requireOwnershipOrAdmin(currentUser, meeting.organizer.id)
        
        val user = userRepository.findById(dto.userId)
            .orElseThrow { NoSuchElementException("User with id ${dto.userId} not found") }
        
        if (meeting.organizer.id == dto.userId) {
            throw IllegalArgumentException("Organizer cannot be invited to their own meeting")
        }
        
        val existingInvitations = invitationRepository.findByMeetingIdAndUserId(dto.meetingId, dto.userId)
        if (existingInvitations.isNotEmpty()) {
            throw IllegalArgumentException("Invitation for user ${dto.userId} to meeting ${dto.meetingId} already exists")
        }
        
        val invitation = Invitation(
            meeting = meeting,
            user = user,
            status = InvitationStatus.PENDING
        )
        return invitationRepository.save(invitation).toResponseDto()
    }

    fun updateInvitation(id: Long, dto: InvitationUpdateDto): InvitationResponseDto {
        val currentUser = SecurityUtils.getCurrentUser(userRepository)
        val invitation = invitationRepository.findById(id)
            .orElseThrow { NoSuchElementException("Invitation with id $id not found") }
        
        SecurityUtils.requireOwnershipOrAdmin(currentUser, invitation.user.id)
        
        invitation.status = dto.status
        
        return invitationRepository.save(invitation).toResponseDto()
    }

    fun deleteInvitation(id: Long) {
        if (!invitationRepository.existsById(id)) {
            throw NoSuchElementException("Invitation with id $id not found")
        }
        invitationRepository.deleteById(id)
    }

    private fun Invitation.toResponseDto() = InvitationResponseDto(
        id = this.id,
        meetingId = this.meeting.id,
        userId = this.user.id,
        status = this.status,
        createdAt = this.createdAt,
        updatedAt = this.updatedAt
    )
}
